optuna/_optimize.py

```diff
@@ -199,7 +199,12 @@ def _run_trial(
     thread: Optional[Thread] = None
 
     if study._storage.is_heartbeat_enabled():
-        study._storage.fail_stale_trials(study._study_id)
+        failed_trial_ids = study._storage.fail_stale_trials(study._study_id)
+        failed_trial_callback = study._storage.get_failed_trial_callback()
+        if failed_trial_callback is not None:
+            for trial_id in failed_trial_ids:
+                failed_trial = copy.deepcopy(study._storage.get_trial(trial_id))
+                failed_trial_callback(study, failed_trial)
```

Comment:
Do you think if it is possible to call the callback before creating a trial instance?
The example code is as follows:

```python
    if study._storage.is_heartbeat_enabled():
        failed_trial_ids = study._storage.fail_stale_trials(study._study_id)
        failed_trial_callback = study._storage.get_failed_trial_callback()
        if failed_trial_callback is not None:
            for trial_id in failed_trial_ids:
                failed_trial = copy.deepcopy(study._storage.get_trial(trial_id))
                failed_trial_callback(study, failed_trial)

    trial = study.ask()

    if study._storage.is_heartbeat_enabled():
        stop_event = Event()
        thread = Thread(
            target=_record_heartbeat, args=(trial._trial_id, study._storage, stop_event)
        )
        thread.start()
```

When I tried to enqueue failed params in the callback function, I need to evaluate two trials because the `_pop_waiting_trial_id` method has been called in the `study.ask()` at L190.

```python
import time
import optuna


def objective(trial):
    return trial.suggest_float("x", 0, 1)


def enqueue_callback(study, trial):
    study.enqueue_trial(trial.params)
    print(f"enqueue {trial.params}")
    return


storage = optuna.storages.RDBStorage(
    url="sqlite:///tmp.db", heartbeat_interval=1, failed_trial_callback=enqueue_callback
)
study = optuna.create_study(storage=storage)

trial = study.ask()
trial.suggest_float("x", 0, 1)
storage.record_heartbeat(trial._trial_id)
time.sleep(2 + 1)

study.optimize(objective, n_trials=2)
```

```console
[I 2021-02-16 13:56:02,983] A new study created in RDB with name: no-name-79bf13ac-2dbf-4ca2-ab16-771f74c6589f
enqueue {'x': 0.06873043341007035}
[I 2021-02-16 13:56:07,079] Trial 1 finished with value: 0.7949246093710458 and parameters: {'x': 0.7949246093710458}. Best is trial 1 with value: 0.7949246093710458.
[I 2021-02-16 13:56:08,122] Trial 2 finished with value: 0.06873043341007035 and parameters: {'x': 0.06873043341007035}. Best is trial 2 with value: 0.06873043341007035.
```

After changing the order of `study.ask()`, the enqueued params were evaluated in the current trial and no need to evaluate two trials.

```console
[I 2021-02-16 13:58:47,844] A new study created in RDB with name: no-name-b125f755-11a5-41a9-bb13-7e0b7fe28960
enqueue {'x': 0.2664680495440712}
[I 2021-02-16 13:58:51,940] Trial 1 finished with value: 0.2664680495440712 and parameters: {'x': 0.2664680495440712}. Best is trial 1 with value: 0.2664680495440712.
```

Reply:
Looks reasonable to me. Fixed.

Reply:
Thank you for your swift action!